src/common/metrics/ngx_wa_metrics.c

```diff
@@ -0,0 +1,490 @@
+    {
+        metrics->shm_zone->noreuse = 1;
+    }
+
+    return NGX_OK;
+}
+
+
+ngx_int_t
+ngx_wa_metrics_init(ngx_wa_metrics_t *metrics, ngx_cycle_t *cycle)
+{
+    ngx_int_t           rc;
+    ngx_wasm_shm_kv_t  *old_shm_kv;
+
+    if (metrics->old_metrics && !metrics->shm_zone->noreuse) {
+        /* reuse old kv store */
+        metrics->shm->data = metrics->old_metrics->shm->data;
+
+        return NGX_OK;
+    }
+
+    rc = ngx_wasm_shm_kv_init(metrics->shm);
+    if (rc != NGX_OK) {
+        return rc;
+    }
+
+    if (metrics->old_metrics && metrics->shm_zone->noreuse) {
+        old_shm_kv = ngx_wasm_shm_get_kv(metrics->old_metrics->shm);
+
+        return metrics_reallocate(metrics, old_shm_kv->rbtree.root,
+                                  old_shm_kv->rbtree.sentinel);
+    }
+
+    return NGX_OK;
+}
+
+
+ngx_int_t
+ngx_wa_metrics_add(ngx_wa_metrics_t *metrics, ngx_str_t *name,
+    ngx_wa_metric_type_e type, uint32_t *out)
+{
+    ssize_t           size = sizeof(ngx_wa_metric_t)
+                             + sizeof(ngx_wa_metric_val_t) * metrics->workers;
+    uint32_t          cas, mid;
+    ngx_int_t         rc, written;
+    ngx_str_t        *p, val;
+    ngx_wa_metric_t  *m;
+    u_char            buf[size];
+
+    if (type != NGX_WA_METRIC_COUNTER
+        && type != NGX_WA_METRIC_GAUGE
+        && type != NGX_WA_METRIC_HISTOGRAM)
+    {
+        return NGX_ERROR;
+    }
+
+    if (name->len > metrics->config.max_metric_name_length) {
+        return NGX_ERROR;
+    }
+
+    mid = ngx_crc32_long(name->data, name->len);
+
+    ngx_wasm_shm_lock(metrics->shm);
+
+    rc = ngx_wasm_shm_kv_get_locked(metrics->shm, NULL, &mid, &p, &cas);
+    if (rc == NGX_OK) {
+        goto done;
+    }
+
+    dd("adding new metric");
+
+    ngx_memzero(buf, size);
+    m = (ngx_wa_metric_t *) buf;
+    m->type = type;
+
+    if (type == NGX_WA_METRIC_HISTOGRAM) {
+        rc = ngx_wa_metrics_histogram_add_locked(metrics, m);
+        if (rc != NGX_OK) {
+            goto error;
+        }
+    }
+
+    val.len = size;
+    val.data = buf;
+
+    rc = ngx_wasm_shm_kv_set_locked(metrics->shm, name, &val, 0, &written);
+
+    if (rc != NGX_OK) {
+        goto error;
+    }
+
+done:
+
+    *out = mid;
+
+error:
+
+    ngx_wasm_shm_unlock(metrics->shm);
+
+    if (rc == NGX_OK) {
+        ngx_wasm_log_error(NGX_LOG_INFO, metrics->shm->log, 0,
+                           "defined %V \"%V\" with id %uD",
+                           metric_type_name(type), name, mid);
+    }
+
+    return rc;
+}
+
+
+ngx_int_t
+ngx_wa_metrics_get(ngx_wa_metrics_t *metrics, uint32_t mid, ngx_uint_t *out)
+{
+    uint32_t          cas;
+    ngx_int_t         rc;
+    ngx_str_t        *n;
+    ngx_wa_metric_t  *m;
+
+    rc = ngx_wasm_shm_kv_get_locked(metrics->shm, NULL, &mid, &n, &cas);
+    if (rc != NGX_OK) {
+        ngx_wasm_log_error(NGX_LOG_ERR, metrics->shm->log, 0,
+                           "metric \"%uD\" not found", mid);
```

Comment:
Would it be possible to return a status code like `NGX_DECLINED` for example, and let the caller decide what errors should be logged? It seems like a very low-level error log to me. I've noticed it from the tests that assert it because they were not testing the log level, and I was wondering where the log was coming from. Since it had no log level test, I thought it was from the test crate, but I was surprised to find it in this low-level ABI.

Reply:
After reading your comment below also regarding the errors logged here, I noticed we can return `NGX_DECLINED` when a metric hasn't been found and `NGX_ERROR` for the other misuses (incrementing a non-counter or recording on a counter) so that a trap message can contain the reason why an operation failed.

Reply:
I think `NGX_ABORT` would be a good rc for "bad usage" (we already use it like this in other ABIs). This way the caller can decide what to do in this case as well. I also like the warning in the low-level metrics ABI for such errors.